src/jupyterlabcontroller/services/lab.py

```diff
@@ -69,175 +206,339 @@ async def create_lab(
         ----------
-        user
-            Identity information for the user spawning the lab.
-        lab
-            Specification for the lab environment to create.
-        image
-            Docker image to run as the lab.
-        delete_first
-            Whether there is an existing lab that needs to be deleted first.
-
-        Returns
-        -------
-        str
-            Cluster-internal URL at which the lab will be listening once it
-            has finished starting.
+        username
+            Username whose environment should be deleted.
 
         Raises
         ------
-        KubernetesError
-            Raised if there is some failure in a Kubernetes API call.
-        MissingSecretError
-            Raised if a secret does not exist.
+        LabDeletionError
+            Raised if the lab deletion failed for any other reason.
+        OperationConflictError
+            Raised if another operation is already in progress.
+        UnknownUserError
+            Raised if no lab currently exists for this user.
         """
-        username = user.username
+        if username not in self._labs:
+            raise UnknownUserError(f"Unknown user {username}")
+        lab = self._labs[username]
+        if not lab.state:
+            raise UnknownUserError(f"Unknown user {username}")
 
-        # Delete an existing failed lab first if needed.
-        if delete_first:
-            self._logger.info("Deleting existing failed lab")
-            await self._lab_state.publish_event(
-                username, f"Deleting existing failed lab for {username}", 2
+        # There are three possible cases.
+        #
+        # 1. A deletion is already in progress, in which case we want to wait
+        #    on it with everyone else who is monitoring the deletion.
+        # 2. No operation is in progress, in which case we start the deletion
+        #    and take responsibility for clearing the lab state when it
+        #    finishes.
+        # 3. A different operation (presumably a spawn) is in progress, in
+        #    which case we want to raise an error, since currently we don't
+        #    support aborting a spawn in progress by deleting the lab.
+        if lab.monitor.in_progress == _LabOperation.DELETE:
+            await lab.monitor.wait()
+        elif not lab.monitor.in_progress:
+            self._builder.build_object_names(username)
+            lab.events.clear()
+            lab.state.status = LabStatus.TERMINATING
+            lab.state.internal_url = None
+            deleter = self._delete_lab(username, lab.state, lab.events)
+            operation = _Operation(
+                _LabOperation.DELETE, deleter, lab.state, lab.events
             )
-            await self._delete_lab_and_namespace(username, 5, 20)
+            await lab.monitor.monitor(operation)
+            await lab.monitor.wait()
+            if lab.state.status == LabStatus.TERMINATED:
+                lab.state = None
```

Comment:
Ah, OK, so maybe we don't want to check for TERMINATED above since we just remove the state entirely.